### PR TITLE
added validation and testing during training 

### PR DESCRIPTION
**Pull request by MaddyThakker**
added params to perform evaluation on validation and test datasets to get an idea of how the model performs on those datasets while training.

### CODE REVIEW

**Review by justinshenk — APPROVED** [review]
Lgtm

File: traja/models/train.py [hunk 1]
```diff
@@ -11,7 +11,7 @@ class HybridTrainer(object):
     """
     Wrapper for training and testing the LSTM model
     Args:
-        model_type: Type of model should be "LSTM"
+        model_type: Type of model should be 'LSTM'
```
Comment (WolfByttner) [inline traja/models/train.py:14]: `model_type` is injected through the model itself and is not needed here. Also, the corresponding __init__ variable doesn't exist.

File: traja/models/train.py [hunk 2]
```diff
@@ -132,7 +132,7 @@ def __str__(self):
         return f"Training model type {self.model_type}"
 
     def fit(
-        self, dataloaders, model_save_path=None, training_mode="forecasting", epochs=50
+            self, dataloaders, model_save_path=None, training_mode="forecasting", epochs=50, validate_every=3, test_every=10
```
Comment (WolfByttner) [inline traja/models/train.py:135]: The new arguments should be described in the function docstring. Also, we should allow `validate_every=None` to disable training-time validation altogether.

File: traja/models/train.py [hunk 3]
```diff
@@ -315,21 +328,21 @@ def fit(
                                 regressor_out, parameters
                             )
 
-                test_loss_forecasting /= len(test_loader.dataset)
+                test_loss_forecasting /= len(data_loader_to_evaluate.dataset)
                 print(
-                    f"====> Mean test set forecasting loss: {test_loss_forecasting:.4f}"
+                    f"====> Mean {current_set} set forecasting loss: {test_loss_forecasting:.4f}"
```
Comment (WolfByttner) [inline traja/models/train.py:333]: You did not update the name of `test_loss_forecasting` to reflect its new usage.

File: traja/models/train.py [hunk 4]
```diff
@@ -315,21 +328,21 @@ def fit(
...
                 )
                 if self.classify:
                     accuracy = correct / total
                     if test_loss_classification != 0:
-                        test_loss_classification /= len(test_loader.dataset)
+                        test_loss_classification /= len(data_loader_to_evaluate.dataset)
                         print(
-                            f"====> Mean test set classifier loss: {test_loss_classification:.4f}; accuracy: {accuracy:.2f}"
+                            f"====> Mean {current_set} set classifier loss: {test_loss_classification:.4f}; accuracy: {accuracy:.2f}"
```
Comment (WolfByttner) [inline traja/models/train.py:340]: You did not update the name of `test_loss_classification` to reflect its new usage.

File: traja/models/train.py [hunk 5]
```diff
@@ -315,21 +328,21 @@ def fit(
...
                         )
 
                 if self.regress:
                     print(
-                        f"====> Mean test set regressor loss: {test_loss_regression:.4f}"
+                        f"====> Mean {current_set} set regressor loss: {test_loss_regression:.4f}"
```
Comment (WolfByttner) [inline traja/models/train.py:345]: You did not update the name of `test_loss_regression` to reflect its new usage.

File: traja/models/train.py [hunk 6]
```diff
@@ -245,15 +246,27 @@ def fit(
                     )
                 )
 
-            # Testing
-            if epoch % 10 == 9 and epoch != 0:
+            # Testing & Validation
+            evaluate_for_this_epoch = False
+            data_loader_to_evaluate = validation_loader
+            current_set = "Validation"
+            if epoch % validate_every == validate_every - 1 and epoch != 0:
```
Comment (WolfByttner) [inline traja/models/train.py:253]: What happens if you should both test and validate? To me it looks like the code would only test.

File: traja/models/train.py [hunk 7]
```diff
@@ -245,15 +246,27 @@ def fit(
...
+                data_loader_to_evaluate = validation_loader
+                evaluate_for_this_epoch = True
+                current_set = "Validation"
+            if epoch % test_every == test_every - 1 and epoch != 0:
```
Comment (WolfByttner) [inline traja/models/train.py:257]: You don't have to submit a validation set. With the current defaults one would trigger a crash. There should be a check to make sure there is a validation set present.

**Review by WolfByttner — CHANGES_REQUESTED** [review]

File: traja/models/train.py (hunk 1 above)
Comment (MaddyThakker) [reply traja/models/train.py:14]: This has always been there in `train.py`
All I did was change `"` -> `'` for consistency.

File: traja/models/train.py (hunk 2 above)
Comment (MaddyThakker) [reply traja/models/train.py:135]: Done.

File: traja/models/train.py (hunk 3 above)
Comment (MaddyThakker) [reply traja/models/train.py:333]: Done.

File: traja/models/train.py (hunk 4 above)
Comment (MaddyThakker) [reply traja/models/train.py:340]: Done.

File: traja/models/train.py (hunk 6 above)
Comment (MaddyThakker) [reply traja/models/train.py:253]: If a particular epoch happens to be a common multiple of `validate_every` and `test_every`, we only test for that epoch. Same is mentioned in the comments now.

File: traja/models/train.py (hunk 7 above)
Comment (MaddyThakker) [reply traja/models/train.py:257]: Done

File: traja/models/train.py (hunk 5 above)
Comment (WolfByttner) [reply traja/models/train.py]: The variable name also needs to be updated

File: traja/models/train.py (hunk 1 above)
Comment (WolfByttner) [reply traja/models/train.py:14]: Please remove the argument then, since it isn't being used

**Comment by WolfByttner** [comment]
There are several lines that are not being tested. Would it be possible to write a test that showcases the new functionality?

File: traja/models/train.py (hunk 1 above)
Comment (MaddyThakker) [reply traja/models/train.py:14]: Done.

File: traja/models/train.py (hunk 5 above)
Comment (MaddyThakker) [reply traja/models/train.py]: Done.

**Comment by MaddyThakker** [comment]
> There are several lines that are not being tested. Would it be possible to write a test that showcases the new functionality?

@WolfByttner Done. Review again. Thanks.

**Comment by WolfByttner** [comment]
`validate_every=None` and `test_every=None` are not tested. I know it isn't much, but I don't want functionality like that to regress at some later stage.

File: traja/models/train.py [hunk 8]
```diff
@@ -159,12 +161,16 @@ def fit(
 
         train_loader = dataloaders["train_loader"]
         test_loader = dataloaders["test_loader"]
+        if 'validation_loader' in dataloaders:
+            validation_loader = dataloaders['validation_loader']
+        else:
+            validate_every = None
```
Comment (WolfByttner) [inline traja/models/train.py:167]: Could you add a test for the "validation data not present" situation, please? That way, no one can remove the functionality without noticing in the future.

File: traja/tests/test_models.py [hunk 9]
```diff
@@ -59,7 +59,7 @@ def test_aevae_jaguar():
     trainer = HybridTrainer(model=model, optimizer_type="Adam", loss_type="huber")
 
     # Train the model
-    trainer.fit(data_loaders, model_save_path, epochs=10, training_mode="forecasting")
+    trainer.fit(data_loaders, model_save_path, epochs=10, training_mode="forecasting", validate_every=5, test_every=10)
```
Comment (WolfByttner) [inline traja/tests/test_models.py:62]: Could we test `validate_every=None` and `test_every=None` as well, please?

**Review by WolfByttner — CHANGES_REQUESTED** [review]

File: traja/tests/test_models.py (hunk 9 above)
Comment (MaddyThakker) [reply traja/tests/test_models.py:62]: Done.

File: traja/models/train.py (hunk 8 above)
Comment (MaddyThakker) [reply traja/models/train.py:167]: What do we want to do if validation data is not present? We shouldn't throw an error sometimes it is ok to not have a validation set.

Comment (WolfByttner) [reply traja/models/train.py:167]: Print nothing about validation, is what I'd expect.

**Review by WolfByttner — APPROVED** [review]

**Review by Saran-nns — APPROVED** [review]
looks good
